Convert shared/constants/metametrics to TS

mcmire opened the pull request:
## Explanation

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

We want to convert NetworkController to TypeScript in order to be able to compare differences in the controller between in this repo and the core repo. To do this, however, we need to convert the dependencies of the controller to TypeScript.

As a part of this effort, this commit converts `shared/constants/metametrics` to TypeScript. Note that simple objects have been largely replaced with enums. There are some cases where I even split up some of these objects into multiple enums.

Progresses #18285.

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

This commit should not introduce any functional changes. All tests should pass, and all functionality in the extension should continue to work.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

mcmire committed:
Convert shared/constants/metametrics to TS

We want to convert NetworkController to TypeScript in order to be able
to compare differences in the controller between in this repo and the
core repo. To do this, however, we need to convert the dependencies of
the controller to TypeScript.

As a part of this effort, this commit converts
`shared/constants/metametrics` to TypeScript. Note that simple objects
have been largely replaced with enums. There are some cases where I even
split up some of these objects into multiple enums.

mcmire commented on shared/constants/metametrics.ts:
```diff
@@ -0,0 +1,637 @@
+  OnboardingTwitterClick = 'External Link Clicked',
+}
+
+export enum MetaMetricsEventAccountType {
```
Note that previously all of these event-related enums were packaged into one object. Enum values must be simple, so we can't nest these into a larger enum.

mcmire commented on shared/constants/metametrics.ts:
```diff
@@ -0,0 +1,637 @@
+  TokenMenu = 'token_menu',
+}
+
+export enum MetaMetricsEventUiCustomizationPropertyValue {
```
Previously this enum was nested in another object (`METAMETRIC_KEY_OPTIONS`). Considering that this object was being used for only one thing, I didn't think it was worth keeping that structure. Also, as above, enum values must be simple so this enum follows the same pattern as the enums above.

mcmire commented on shared/constants/metametrics.ts:
```diff
@@ -0,0 +1,637 @@
+/**
+ * Used to identify events that are triggered by the background process.
+ */
+export const METAMETRICS_BACKGROUND_PAGE_OBJECT: MetaMetricsPageObject = {
```
Once we upgrade TypeScript we should be able to use `satisfies MetaMetricsPageObject` instead of typing it like this, but this should work for the time being.

mcmire commented on app/scripts/background.js:
```diff
@@ -24,11 +24,11 @@ import {
-  EVENT,
-  EVENT_NAMES,
-  TRAITS,
+  REJECT_NOTIFICATION_CLOSE,
```
Fixed this typo while I was at it.

mcmire committed:
Merge branch 'develop' into convert-metametrics-constants

mcmire committed:
Correct MetaMetricsContext

mcmire committed:
Capitalize this

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Fix JSDoc

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Add missing JSDoc

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Add missing JSDoc.

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Add missing JSDoc.

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Use LedgerTransportTypes

mcmire commented on shared/constants/metametrics.ts:
```diff
@@ -0,0 +1,637 @@
+  /**
+   * The type of ledger connection set by user preference.
+   */
+  ledger_connection_type?: 'ledgerLive' | 'webhid' | 'u2f';
```
Ah yes! I didn't know this existed. Updated in [150a143](https://github.com/MetaMask/metamask-extension/pull/18353/commits/150a1437113799742b1c4d10605a3dd971007643).

mcmire commented on shared/constants/metametrics.ts:
```diff
@@ -0,0 +1,637 @@
+    /**
+     * The user agent of the application.
+     */
+    userAgent: string;
```
Good catch! Fixed in [3f7711b](https://github.com/MetaMask/metamask-extension/pull/18353/commits/3f7711bdb60f8f37bb1fc340d4d02f98333edf0f).

mcmire committed:
Convert more instances of EVENT_NAMES, etc.

Gudahtt commented on ui/components/app/detected-token/detected-token.js:
```diff
@@ -86,11 +91,11 @@ const DetectedToken = ({ setShowDetectedTokens }) => {
       sensitiveProperties: {
         tokens: tokensDetailsList,
-        location: EVENT.LOCATION.TOKEN_DETECTION,
+        location: MetaMetricsEventLocation,
```
Looks like the property was forgotten here:
```suggestion
        location: MetaMetricsEventLocation.TokenDetection,
```

Gudahtt commented:
Looks great! I can approve after the one problem I found is resolved

mcmire committed:
Fix detected-token.js

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Merge branch 'develop' into convert-metametrics-constants

mcmire committed:
Normalize constants

- There are several categories of event sources. Rearrange the constant
  names to accentuate this. For instance, `MetaMetricsEventNetworkSource`
  would seem (when reading this as English) to indicate that we have a
  "network source" that is related to a MetaMetrics event. But really,
  it's a MetaMetrics event source that we've classified under "network".
- An external link type is just a link type (according to how it's
  used). There is no other kind of link type. Rename this constant
  accordingly.
- We don't include "properties" in the name of the types. For instance a
  key type is passed as a property, but we just call the type
  `MetaMetricsEventKeyType`. Rename
  `MetaMetricsEventUiCustomizationPropertyValue` accordingly.

Gudahtt reviewed (dismissed):
LGTM!

mcmire committed:
Rename lingering instance

mcmire committed:
Merge branch 'develop' into convert-metametrics-constants

mcmire committed:
Fix another instance of the old constant sneaking in

mcmire commented on ui/store/action-queue/index.ts:
```diff
@@ -192,8 +195,8 @@ async function processActionRetryQueue() {
   try {
     if (actionRetryQueue.length > 0) {
       const metametricsPayload = {
-        category: EVENT.SOURCE.SERVICE_WORKERS,
```
I went ahead and fixed this instance since it was used in two places: [83b3f0a](https://github.com/MetaMask/metamask-extension/pull/18353/commits/83b3f0af9bcb621b7921f087352bf9fb34affaf4)

Gudahtt replied:
Makes sense!

mcmire committed:
Merge branch 'develop' into convert-metametrics-constants

mcmire commented:
CI looked like it was about to pass before I merged `develop` again, so I don't anticipate any more changes.

mcmire committed:
Merge branch 'develop' into convert-metametrics-constants

brad-decker reviewed (approved):
LGTM, thanks @mcmire
